src/states/node/mod.rs

```diff
@@ -550,15 +550,13 @@ impl Node {
             }
         } else {
             // Remove any untrusted trailing section infos.
-            // TODO: remove wasted clone. Only useful when msg isnt trusted for log msg.
-            let msg_clone = signed_msg.clone();
             while match signed_msg.previous_hop() {
                 None => true,
                 Some(hop) => !self.is_trusted(hop)?,
             } {
                 // We don't know the last hop! Try the one before that.
                 if !signed_msg.pop_previous_hop() {
-                    debug!("{} Untrusted message: {:?}", self, msg_clone);
+                    debug!("{} Untrusted message: {:?}", self, signed_msg);
```

Comment:
hmm I think the TODO isn't maybe too clear. by just using signed_msg here instead of msg_clone, we'd only be logging the message without any proving_sections in it I think as the `pop_previous_hop` calls through the while loop would be stripping them out. So the point of the cloned object before starting the while loop was to keep a copy of the full signed_msg(with proving_sections) for the log here which ofc can be done better by potentially not popping out invalid ones if we can check is_trusted across the list directly and then print this message and then do the pop out...

Reply:
Thank you for the feedback! I missed the mutation on `signed_msg` between the clone() and log operations. I'm taking your suggestion to run the `is_trusted` check over the list and logging the message before popping the untrusted hops.
However, while doing so, I noticed that the existing implementation does not error out if the `source_section` field in the message fails the `is_trusted` check, the code only requires that this field is set. Is the source section always trusted?

Reply:
no problem, although this does add a bit of complexity to a portion of code thats going to get deprecated with the secure msg delivery RFCs coming up. So for now I'd probably just vote to close this PR in favor of this TODO getting deprecated soon enough.

Reply:
That sounds fair, closing.